lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,122 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
 
         def serializable_hash(options = nil)
           options ||= {}
+          serialized_hash = {}
+
           if serializer.respond_to?(:each)
-            result = serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
+            serialized_hash = serialize_array_without_root(serializer, options)
           else
-            hash = {}
+            serialized_hash = resource_object_for(serializer)
+            serialized_hash = add_resource_relationships(serialized_hash, serializer)
```

Comment:
This `x = f(x, y)` still seems a bit shady to me. I think it would be best to make `add_resource_relationships` modify `serialized_hash` in place. What do you think?